### ISSUE TITLE
Unable to Install swirl for R3.1.2

### ISSUE THREAD

**Issue by apekshaguptaa** [swirldev/swirl#251]
Please find below the log for the same :

> install.packages("swirl")
Installing package into ‘D:/Profiles/apGupta/Documents/R/win-library/3.1’
(as ‘lib’ is unspecified)
--- Please select a CRAN mirror for use in this session ---
Warning: unable to access index for repository http://ftp.iitm.ac.in/cran/bin/windows/contrib/3.1
Warning: unable to access index for repository http://www.stats.ox.ac.uk/pub/RWin/bin/windows/contrib/3.1
Warning messages:
1: In open.connection(con, "r") :
  unable to connect to 'cran.r-project.org' on port 80.
2: package ‘swirl’ is not available (for R version 3.1.2)

**Comment by WilCrofter** [comment]
The diagnostic indicates trouble with a connection to the CRAN mirrors referenced. You can visit these links in a browser. Doing so, the first, at iitm.ac.in, looks fine to me; it has the current version of swirl. The server may have been down or busy when you tried. Just trying again might work.

The second mirror referenced, at ox.ac.uk, contains just a few packages and swirl is not among them, so I would avoid that one if you can.

The documentation for `install.packages` might be useful. Just type `help(install.packages)` in the R console. The "universal" mirror, http://cran.rstudio.com, is generally reliable.

**Comment by dnduati** [comment]
I use windows 7 and R version 3.1.2.
I am getting the following errors:

1.) Warning in install.packages :
  unable to access index for repository http://www.stats.ox.ac.uk/pub/RWin/bin/windows/contrib/3.1
Installing package into ‘C:/Users/Dorcas/Documents/R/win-library/3.1’
(as ‘lib’ is unspecified)

2.) Warning in install.packages :
  cannot open: HTTP status was '407 Proxy Authentication Required'

3.) > library("swirl")
Error in library("swirl") : there is no package called ‘swirl’
> swirl()
Error: could not find function "swirl"

Please assist so that I can do some assignment on swirl here at work. At home I have no problem my laptop works very well.

**Comment by dnduati** [comment]
Swirl package is working now but R-programming assignments I have tried to fork and am getting the following error:

![capture](https://cloud.githubusercontent.com/assets/7793349/5974839/e25d9148-a889-11e4-9b43-a0f28cfd3391.PNG)

**Comment by ivancope** [comment]
![swirl failed to install r3 1 2 on windows pc](https://cloud.githubusercontent.com/assets/10506347/6028412/043d3d3e-abf2-11e4-9eb1-2899565a4504.jpg)
With R3.1.2 on aWindows PC:
 I tried  install.packages("swirl"). It looked OK except for warning message but failed after library("swirl") command. 
Perhaps the attached screenshot tells more than i can.
Thanks for any help.

**Comment by WilCrofter** [comment]
@ivancope, for some reason package yaml, a swirl dependency, failed to install. There was a warning that, though the yaml package was successfully downloaded and unpacked to a temporary location, R was unable to move it to your library folder. (Permission problem?) Try installing it independently,
```
install.packages("yaml")
```
If that doesn't work try installing from a manually downloaded tar.gz file. If all else fails, I've successfully moved packages from temporary installation locations to library directories by hand.

**Comment by dnduati** [comment]
I will show us my error from my desktop

Install.packages("swirl")
library(swirl)
swirl()

Thanks

**Comment by WilCrofter** [comment]
@dnduati, I'm not sure what you are trying to do. If you just want to install and run a swirl course, cloning isn't necessary.  See the help file for InstallCourses:
```
library(swirl)
?InstallCourses
```
If you want to modify swirl_courses, I see you have forked swirldev/swirl_courses, at github, so it's just  a matter of setting up a local git repository. Since you've been using github for a while, I imagine github has your public key, so I'm not sure why the clone is complaining about access. Is your key pair on the machine you are using locally?

**Comment by dnduati** [comment]
Thank you  l will read ?installcourses

**Comment by dnduati** [comment]
I have tried what you told me and got the following error.

> ?install.courses
No documentation for 'install.courses' in specified packages and libraries:
you could try '??install.courses'
> ??install.courses
>
Search Results


The search string was "install.courses"
No results found

I didn't manage to install swirl packages

On Wed, Feb 4, 2015 at 9:01 AM, Dorcas Nduati <dnduati@gmail.com> wrote:

> Thank you  l will read ?installcourses
> On 4 Feb 2015 08:44, "William R Bauer" <notifications@github.com> wrote:
>
>> @dnduati <https://github.com/dnduati>, I'm not sure what you are trying
>> to do. If you just want to install and run a swirl course, cloning isn't
>> necessary. See the help file for InstallCourses:
>>
>> library(swirl)
>> ?InstallCourses
>>
>> If you want to modify swirl_courses, I see you have forked
>> swirldev/swirl_courses, at github, so it's just a matter of setting up a
>> local git repository. Since you've been using github for a while, I imagine
>> github has your public key, so I'm not sure why the clone is complaining
>> about access. Is your key pair on the machine you are using locally?
>>
>> --
>> Reply to this email directly or view it on GitHub
>> <https://github.com/swirldev/swirl/issues/251#issuecomment-72794252>.
>>
>


-- 
Dorcas Nduati
PHPT Dept.
Room 111
UoN

Office:0202453621 Ext 27107
Skype:dorcas.waithira.nduati

**Comment by ivancope** [comment]
@WilCrofter Thanks for your reply.
I'm still struggling but maybe some light in view:  (Apologies for being so verbose, but i'm not sure what is important.)
1)
I started off trying to install.packages("yam1")  ... [..instead of ("yaml"), but i soon saw that error, highlighted by Warning message:
package ‘yam1’ is not available (for R version 3.1.2) 


2)
Trying install.packages("yaml") gave similar problems to before: 
"  package ‘yaml’ successfully unpacked and MD5 sums checked
Warning: unable to move temporary installation ‘C:\Users\User\Documents\R\win-library\3.1\filebe0258307f\yaml’ to ‘C:\Users\User\Documents\R\win-library\3.1\yaml’

The downloaded binary packages are in
        C:\Users\User\AppData\Local\TempTh\RtmpeIi4y0\downloaded_packages
> library("swirl")
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called ‘yaml’
Error: package or namespace load failed for ‘swirl’   "
3)
So i decided to see what was in the folders:
3a)
C:\Users\User\Documents\R\win-library\3.1\file16e477402bd8 : had a copy of folder yaml from yesterday,3Feb, which contained tests which contains test_yaml_load.R     .   [I opened this into Rstudio. It seemed to be be R source code for various tests.  This was the only filennnnnnnnn folder with yaml]
3b) 
C:\Users\User\AppData\Local\Temp\RtmpeIi4y0\downloaded_packages had yaml_2.1.13.zip              of today 4 Feb
[ Should i try to unzip.  If so how do i do that?]
3c)
C:\Users\User\Documents\R\win-library\3.1 did not contain yaml
-------------
So i tried using Windows Explorer to copy yaml from 3.1\file16e477402bd8 to C:\Users\User\Documents\R\win-library\3.1   in the hope that would fix the problem when i entered library("swirl"), but no go. It still couldn't find yaml:
> library("swirl")
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called ‘yaml’
Error: package or namespace load failed for ‘swirl’

Your comment would be greatly appreciated.

**Comment by WilCrofter** [comment]
@ivancope, my memory is dim but if the installer's only problem is moving
```
temporary installation ‘C:\Users\User\Documents\R\win-library\3.1\filebe0258307f\yaml’ to ‘C:\Users\User\Documents\R\win-library\3.1\yaml’
```
then moving it there by hand, which I think you've done, should solve at least that part of the problem. Maybe restarting the R session is necessary, but I'm not sure why it should be.

I don't find any references to yaml package installation problems on the web, but [here](http://stackoverflow.com/questions/5700505/windows-7-update-packages-problem-unable-to-move-temporary-installation) is a stackoverflow post blaming antivirus software.

If antivirus is preventing the copy, the same thing will probably happen if you manually [download the windows binary](http://cran.r-project.org/bin/windows/contrib/3.1/yaml_2.1.13.zip) and
```
install.packages("path/to/yaml_2.1.13.zip", repos=NULL)
```
You may have done this already.

**Comment by WilCrofter** [comment]
@dnduati, loading the swirl library and including the upper case letters is important. From within an R session in the RStudio console or whatever you are using, you must type precisely this:
```
library(swirl)
?InstallCourses
```
IOW, it must be exactly `?InstallCourses`, not `?install.courses` or `?Installcourses` or other variations.

The easiest way to install the R Programming course is
```
library(swirl)
install_from_swirl("R Programming")
```
Other courses can be installed the same way. Just replace "R Programming" with "Statistical Inference" or whichever course you'd like.

**Comment by dnduati** [comment]
Thank you will try when l reach home

**Comment by all4beals5** [comment]
I have a similar issue when installing swirl. R seems to have installed it correctly, but when I start swirl I get an error message.

![swirl r 1](https://cloud.githubusercontent.com/assets/10562189/6049815/6c22f658-ac80-11e4-8a66-f17f543acca2.png)
![swirl r 2](https://cloud.githubusercontent.com/assets/10562189/6049817/6e630386-ac80-11e4-833e-564ff098b977.png)

**Comment by ivancope** [comment]
@WillCrofter
I'm not sure i understood what you were saying in your last post, but i tried quitting my sessions and tried
library ("swirl")  but it didn't work.
Then i tried:
install.packages("path/to/yaml_2.1.13.zip", repos=NULL), but again no success.
without quotes and with quotes, but again no luck.  (See transcript below.)

Thanks for your help.

> library("swirl")
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called ‘yaml’
Error: package or namespace load failed for ‘swirl’
> install.packages(C:\Users\User\AppData\Local\Temp\yaml_2.1.13.zip, repos=NULL)
Error: unexpected input in "install.packages(C:\"
> install.packages(“C:\Users\User\AppData\Local\Temp\yaml_2.1.13.zip”, repos=NULL)
Error: unexpected input in "install.packages(“"

**Comment by WilCrofter** [comment]
@all4beals5, the diagnostic says that your library location is not writable. This means you can't install packages to it. Presumably you don't have write permission. There are work-arounds. Please read the documentation for `install.packages` using the R help files. e.g., from the R prompt:
```
?install.packages
```

**Comment by WilCrofter** [comment]
@ivancope, sorry if I've confused you. My guess is that anti-virus software is preventing R from moving the yaml package from the temporary location where it was unpacked, to the permanent location where R can load and use it. I may be wrong. I'm just guessing, but since your problem is specific to one package, an anti-virus issue seems plausible.

I'm also guessing that @all4beals5 has a permission problem, which is different, but also prevents a package from being installed in its proper place. This a better bet, since the diagnostic says that `C://Program Files/R/R-3.1.2/library is not writable.` I have no idea why it isn't. 

System problems, which these seem to be, are nearly impossible to diagnose and solve remotely.  Your best bet is probably to read the documentation for `.libPaths` (notice the leading dot) and `install.packages()`. These can, of course, be found in the R help files. `?.libPaths`, `?install.packages`.

[This article at stackoverflow](http://stackoverflow.com/questions/15170399/changing-r-default-library-path-using-libpaths-in-rprofile-site-fails-to-work) may be helpful, especially to @all4beals5.

**Comment by ivancope** [comment]
@WillCrofter
Thanks again.
I will try to read the docs you reference, but ....i thought that i had manually copied files.  

Could you tell me what files R expects in each location relating to yaml, and (if possible) the structure of these files so i can try to check if the files look OK.

**Comment by dnduati** [comment]
> library(swirl) # even if it is library("swirl") bringing the same error.
> install_from_swirl("R Programming")
Error in function (type, msg, asError = TRUE)  :
  Failed to connect to github.com port 80: Connection refused

I think it can be solved by going back to the lower version of R.

The problem is persistent

On Thu, Feb 5, 2015 at 10:36 AM, Ivan Copelowitz <notifications@github.com>
wrote:

> @WillCrofter
> Thanks again.
> I will try to read the docs you reference, but ....i thought that i had
> manually copied files.
>
> Could you tell me what files R expects in each location relating to yaml,
> and (if possible) the structure of these files so i can try to check if the
> files look OK.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/swirldev/swirl/issues/251#issuecomment-73005943>.
>



-- 
Dorcas Nduati
PHPT Dept.
Room 111
UoN

Office:0202453621 Ext 27107
Skype:dorcas.waithira.nduati

**Comment by WilCrofter** [comment]
@dnduati, ` Failed to connect to github.com port 80: Connection refused` indicates that  github.com was unavailable. The usual reason for that is blocking by an institution such as a workplace. I'm not sure why an earlier version of R would help. Is package `httr` installed?

**Comment by all4beals5** [comment]
@WilCrofter yes, that location is not writable for some reason, but then I got asked if I wanted to use a personal library. I said yes and my second screenshot shows successful installation. I'm assuming that it needs to be saved in a specific location (not a personalized location)?

**Comment by WilCrofter** [comment]
@ivancope, here is a directory tree of my installation of R's yaml package. I'm on a linux machine, so my installation will look slightly different than a windows installation. Windows will have dll's while linux will have so's for instance. The yaml subdirectory was in the first location listed by `.libPaths()` on my machine.

yaml
├── CHANGELOG
├── DESCRIPTION
├── help
│   ├── aliases.rds
│   ├── AnIndex
│   ├── paths.rds
│   ├── yaml.rdb
│   └── yaml.rdx
├── html
│   ├── 00Index.html
│   └── R.css
├── implicit.re
├── INDEX
├── libs
│   └── yaml.so
├── LICENSE
├── Meta
│   ├── hsearch.rds
│   ├── links.rds
│   ├── nsInfo.rds
│   ├── package.rds
│   └── Rd.rds
├── NAMESPACE
├── R
│   ├── yaml
│   ├── yaml.rdb
│   └── yaml.rdx
├── tests
│   ├── files
│   │   └── test.yml
│   ├── test_as_yaml.R
│   ├── test_yaml_load_file.R
│   └── test_yaml_load.R
└── THANKS

7 directories, 27 files

**Comment by WilCrofter** [comment]
@all4beals5, does the personal library's location appear in the list given by `.libLocs()`? If not, you must add its location to the list of places R uses to search for installed packages. The documentation for doing so is not all that clear, I'm afraid, but it's all we have.

**Comment by WilCrofter** [comment]
On Thu, 05 Feb 2015 06:13:10 -0800
all4beals5 <notifications@github.com> wrote:

> that location is not writable for some reason

BTW, it would be a good idea to find out why it is not writable. Is there a sysadmin or help desk available to you?

**Comment by all4beals5** [comment]
I closed and restarted R. Then I followed the instructions on the Coursera R course, swirl programming assignment page (https://class.coursera.org/rprog-011/assignment/view?assignment_id=9) and was successful in my installation and launching of swirl. Weird. I didn't do anything differently from my previous attempts only that I didn't have anything else in my command/prompt window.

**Comment by WilCrofter** [comment]
Great! Glad it worked. Cheers.

**Comment by ivancope** [comment]
@WilCrofter 
Your tree so overwhelmed me, i decided to give up!!
However, i was stupid enough to try implementing Swirl after a few days ... today, in fact.

The gods were merciful, or perhaps you offered a pleasing sacrifice on my behalf.

Anyway, FOR SOME REASON KNOWN ONLY TO THE GODS OF COMPUTERS ALONE, the implementation worked without trouble???!!!

(I can only surmise that something was reset somehow.)

i THANK YOU MOST SINCERELY for helping me. It was sometimes a happy journey for me.

Best Wishes.

**Comment by WilCrofter** [comment]
@ivancope 
I've very glad everything worked out. I appreciate your thanks, and hope you enjoy the courses which Nick, Gina, Sean, (and to some extent David Kane and I) managed to churn out last year.

For the immediate future, I'll have to focus on [another exciting and worthwhile project](http://www.hopkinsmedicine.org/news/media/releases/team_will_use_brain_initiative_grant_to_design_next_generation_imaging_technology), leaving everyone in Sean's capable hands, mostly. (Thanks, Sean.)

**Comment by ivancope** [comment]
@WilCrofter
 I certainly wish you success in the project and hope it leads to significant breakthroughs .But i think that even the refinement of the measurement and the data treatment is also likely to lead to greater understanding and hence other exciting studies e.g a detailed analysis of how we think, effect of the subconscious on that and also allied to that, the effect of past traumatic events on our current every day thinking.

We live in exciting times, but the future is even more "WOW!!!".

Thanks again for your time and effort on my behalf.  (So mundane in comparison.)

**Comment by ivancope** [comment]
@WilCrofter.

Just to add my early impression of Swirl courses. I'm finding them very useful - especially where they fill some gaps left by the Coursera lectures.

**Comment by WilCrofter** [comment]
Ivan,

Thanks for your very kind words and wishes. The next-gen brain imaging project is indeed an amazing effort, involving everything from a troupe of baboons to quantum chemistry. The two older PI's, along with Henry Wagner, were pioneers of PET brain imaging in the early 1980's.

Gina and I will both be sticking with swirl. Gina's Exploratory Data Analysis is in beta right now. We will publish a swirl course for people who want to write their own swirl courses within a few days, and Calculus and Linear Algebra for Data Science by summer. 

We'll just be distracted until the brain imaging project gets rolling as smoothly as swirl.

Cheers,
Bill

**Comment by cas56** [comment]
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called ‘RCurl’
Error: package or namespace load failed for ‘swirl’

Having error in loading the library.

**Comment by Fareeha2** [comment]
Hi!
I am also unable to install swirl pakage and getting this error. can anybody please help me?

> install.packages("swirl")
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  unable to access index for repository http://cran.rstudio.com/src/contrib
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  unable to access index for repository http://www.stats.ox.ac.uk/pub/RWin/src/contrib
Installing package into ‘C:/Users/Fareeha Sultan/Documents/R/win-library/3.2’
(as ‘lib’ is unspecified)
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  unable to access index for repository http://cran.rstudio.com/src/contrib
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  unable to access index for repository http://www.stats.ox.ac.uk/pub/RWin/src/contrib
Warning in install.packages :
  package ‘swirl’ is not available (for R version 3.2.0)
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  unable to access index for repository http://cran.rstudio.com/bin/windows/contrib/3.2
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  InternetOpenUrl failed: 'A connection with the server could not be established'
Warning in install.packages :
  unable to access index for repository http://www.stats.ox.ac.uk/pub/RWin/bin/windows/contrib/3.2

**Comment by dnduati** [comment]
I use Windows 7 and r version 3.3.1

Use the fillowing to install swirl:

install.packages("swirl")
library("swirl")

If you want to install one of the courses use:

install_from_swirl("Regression Models")

Regards
